src/gamepad/mapping.vala

```diff
@@ -0,0 +1,234 @@
+// This file is part of GNOME Games. License: GPLv3
+
+private class LibGamepad.DPad : Object {
+	public InputType types[4];
+	public int values[4];
+	public int axis_values[2];
+
+	public DPad () {
+		axis_values[0] = axis_values[1] = 0;
+		types[0] = types[1] = types[2] = types[3] = InputType.INVALID;
+	}
+}
+
+public class LibGamepad.Mapping : Object {
+	private InputType[] buttons_type;
+	private int[] buttons_value;
+	private InputType[] axes_type;
+	private int[] axes_value;
+	private DPad[] dpads;
+
+	public Mapping.from_sdl_string (string? mapping_string) throws MappingError {
+		if (mapping_string == null || mapping_string == "")
+			throw new MappingError.NOT_A_MAPPING ("The mapping string can't be null nor empty.");
+
+		var mappings = mapping_string.split (",");
+		foreach (var mapping in mappings) {
+			if (mapping.split (":").length == 2) {
```

Comment:
We are computing `mapping.split (":")` 3 times, it's probably better to store it and use the result 3 times.